// metal-pagination/src/Pagination.soy.js
/* jshint ignore:start */
import Component from 'metal-component';
import Soy from 'metal-soy';
var templates;
goog.loadModule(function(exports) {

// This file was automatically generated from Pagination.soy.
// Please don't edit this file by hand.

/**
 * @fileoverview Templates in namespace Pagination.
 * @public
 */

goog.module('Pagination.incrementaldom');

/** @suppress {extraRequire} */
var soy = goog.require('soy');
/** @suppress {extraRequire} */
var soydata = goog.require('soydata');
/** @suppress {extraRequire} */
goog.require('goog.i18n.bidi');
/** @suppress {extraRequire} */
goog.require('goog.asserts');
var IncrementalDom = goog.require('incrementaldom');
var ie_open = IncrementalDom.elementOpen;
var ie_close = IncrementalDom.elementClose;
var ie_void = IncrementalDom.elementVoid;
var ie_open_start = IncrementalDom.elementOpenStart;
var ie_open_end = IncrementalDom.elementOpenEnd;
var itext = IncrementalDom.text;
var iattr = IncrementalDom.attr;


/**
 * @param {Object<string, *>=} opt_data
 * @param {(null|undefined)=} opt_ignored
 * @param {Object<string, *>=} opt_ijData
 * @return {void}
 * @suppress {checkTypes}
 */
function $render(opt_data, opt_ignored, opt_ijData) {
  ie_open('ul', null, null,
      'class', 'pagination' + (opt_data.elementClasses ? ' ' + opt_data.elementClasses : ''));
    if (opt_data.showControls == true) {
      var disabled__soy8 = ! opt_data.circular && opt_data.page == 0 ? true : false;
      $renderControlElement_({ariaLabel: opt_data.strings.prevAriaLabel, content: opt_data.strings.prev, disabled: disabled__soy8, index: 0}, null, opt_ijData);
    }
    var iLimit14 = opt_data.total;
    for (var i14 = 0; i14 < iLimit14; i14++) {
      var active__soy15 = opt_data.page == i14 ? true : false;
      $renderElement_({active: active__soy15, content: opt_data.offset + i14, index: i14}, null, opt_ijData);
    }
    if (opt_data.showControls == true) {
      var disabled__soy22 = ! opt_data.circular && opt_data.page == opt_data.total - 1 ? true : false;
      $renderControlElement_({ariaLabel: opt_data.strings.nextAriaLabel, content: opt_data.strings.next, disabled: disabled__soy22, index: 1}, null, opt_ijData);
    }
  ie_close('ul');
}
exports.render = $render;
if (goog.DEBUG) {
  $render.soyTemplateName = 'Pagination.render';
}


/**
 * @param {Object<string, *>=} opt_data
 * @param {(null|undefined)=} opt_ignored
 * @param {Object<string, *>=} opt_ijData
 * @return {void}
 * @suppress {checkTypes}
 */
function $renderElement_(opt_data, opt_ignored, opt_ijData) {
  ie_open_start('li');
      iattr('class', 'pagination-item' + (opt_data.active ? ' active' : ''));
      if (! opt_data.active) {
        iattr('data-onclick', 'onClickItem');
      }
      iattr('data-index', opt_data.index);
  ie_open_end();
    ie_open('a', null, null,
        'href', '#');
      itext((goog.asserts.assert((opt_data.content) != null), opt_data.content));
    ie_close('a');
  ie_close('li');
}
exports.renderElement_ = $renderElement_;
if (goog.DEBUG) {
  $renderElement_.soyTemplateName = 'Pagination.renderElement_';
}


/**
 * @param {Object<string, *>=} opt_data
 * @param {(null|undefined)=} opt_ignored
 * @param {Object<string, *>=} opt_ijData
 * @return {void}
 * @suppress {checkTypes}
 */
function $renderControlElement_(opt_data, opt_ignored, opt_ijData) {
  ie_open_start('li');
      iattr('class', 'pagination-control' + (opt_data.disabled ? ' disabled' : ''));
      if (! opt_data.disabled) {
        iattr('data-onclick', 'onClickControls');
      }
      iattr('data-control-index', opt_data.index);
  ie_open_end();
    ie_open('a', null, null,
        'href', '#',
        'aria-label', opt_data.ariaLabel);
      itext((goog.asserts.assert((opt_data.content) != null), opt_data.content));
    ie_close('a');
  ie_close('li');
}
exports.renderControlElement_ = $renderControlElement_;
if (goog.DEBUG) {
  $renderControlElement_.soyTemplateName = 'Pagination.renderControlElement_';
}

exports.render.params = ["circular","elementClasses","offset","page","strings","showControls","total"];
exports.render.types = {"circular":"any","elementClasses":"any","offset":"any","page":"any","strings":"any","showControls":"any","total":"any"};
exports.renderElement_.params = ["active","content","index"];
exports.renderElement_.types = {"active":"any","content":"any","index":"any"};
exports.renderControlElement_.params = ["ariaLabel","content","disabled","index"];
exports.renderControlElement_.types = {"ariaLabel":"any","content":"any","disabled":"any","index":"any"};
templates = exports;
return exports;

});

class Pagination extends Component {}
Soy.register(Pagination, templates);
export { Pagination, templates };
export default templates;
/* jshint ignore:end */
