// clay-charts/lib/Chart.soy.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.templates = exports.ClayChart = undefined;

var _metalComponent = require('metal-component');

var _metalComponent2 = _interopRequireDefault(_metalComponent);

var _metalSoy = require('metal-soy');

var _metalSoy2 = _interopRequireDefault(_metalSoy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /* jshint ignore:start */


var templates;
goog.loadModule(function (exports) {
  var soy = goog.require('soy');
  var soydata = goog.require('soydata');
  // This file was automatically generated from Chart.soy.
  // Please don't edit this file by hand.

  /**
   * @fileoverview Templates in namespace ClayChart.
   * @public
   */

  goog.module('ClayChart.incrementaldom');

  goog.require('goog.soy.data.SanitizedContent');
  var incrementalDom = goog.require('incrementaldom');
  goog.require('soy.asserts');
  var soyIdom = goog.require('soy.idom');

  /**
   * @param {{
   *  _loading: (boolean|null|undefined),
   *  id: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $render(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    opt_data = opt_data || {};
    /** @type {boolean|null|undefined} */
    var _loading = soy.asserts.assertType(opt_data._loading == null || goog.isBoolean(opt_data._loading) || opt_data._loading === 1 || opt_data._loading === 0, '_loading', opt_data._loading, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var id = soy.asserts.assertType(opt_data.id == null || goog.isString(opt_data.id) || opt_data.id instanceof goog.soy.data.SanitizedContent, 'id', opt_data.id, '!goog.soy.data.SanitizedContent|null|string|undefined');
    var attrs__soy33 = function attrs__soy33() {
      incrementalDom.attr('class', 'metal-chart-container');
      if (id) {
        incrementalDom.attr('id', id);
      }
    };
    incrementalDom.elementOpenStart('div');
    attrs__soy33();
    incrementalDom.elementOpenEnd();
    var chartAttrs__soy44 = function chartAttrs__soy44() {
      if (_loading) {
        incrementalDom.attr('hidden', '');
      }
      incrementalDom.attr('ref', 'chart');
    };
    incrementalDom.elementOpenStart('div');
    incrementalDom.attr('class', 'metal-chart');
    chartAttrs__soy44();
    incrementalDom.elementOpenEnd();
    incrementalDom.elementClose('div');
    var placeholderAttrs__soy53 = function placeholderAttrs__soy53() {
      if (!_loading) {
        incrementalDom.attr('hidden', '');
      }
      incrementalDom.attr('ref', 'placeholder');
    };
    incrementalDom.elementOpenStart('div');
    placeholderAttrs__soy53();
    incrementalDom.elementOpenEnd();
    $loading(null, null, opt_ijData);
    incrementalDom.elementClose('div');
    incrementalDom.elementClose('div');
  }
  exports.render = $render;
  /**
   * @typedef {{
   *  _loading: (boolean|null|undefined),
   *  id: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }}
   */
  $render.Params;
  if (goog.DEBUG) {
    $render.soyTemplateName = 'ClayChart.render';
  }

  /**
   * @param {Object<string, *>=} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $loading(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    incrementalDom.elementOpenStart('div');
    incrementalDom.attr('aria-hidden', 'true');
    incrementalDom.attr('class', 'loading-icon');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementOpenStart('span');
    incrementalDom.attr('class', 'loading-animation');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementClose('span');
    incrementalDom.elementClose('div');
  }
  exports.loading = $loading;
  if (goog.DEBUG) {
    $loading.soyTemplateName = 'ClayChart.loading';
  }

  exports.render.params = ["_loading", "id"];
  exports.render.types = { "_loading": "bool", "id": "string" };
  exports.loading.params = [];
  exports.loading.types = {};
  exports.templates = templates = exports;
  return exports;
});

var ClayChart = function (_Component) {
  _inherits(ClayChart, _Component);

  function ClayChart() {
    _classCallCheck(this, ClayChart);

    return _possibleConstructorReturn(this, (ClayChart.__proto__ || Object.getPrototypeOf(ClayChart)).apply(this, arguments));
  }

  return ClayChart;
}(_metalComponent2.default);

_metalSoy2.default.register(ClayChart, templates);
exports.ClayChart = ClayChart;
exports.templates = templates;
exports.default = templates;
/* jshint ignore:end */
