// metal-jsx/lib/JSXComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

require('./iDOMHelpers');

var _metal = require('metal');

var _metalComponent = require('metal-component');

var _metalComponent2 = _interopRequireDefault(_metalComponent);

var _metalIncrementalDom = require('metal-incremental-dom');

var _metalIncrementalDom2 = _interopRequireDefault(_metalIncrementalDom);

var _JSXDataManager = require('./JSXDataManager');

var _JSXDataManager2 = _interopRequireDefault(_JSXDataManager);

var _JSXRenderer = require('./JSXRenderer');

var _JSXRenderer2 = _interopRequireDefault(_JSXRenderer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * A component that has built-in integration with JSX templates. Example:
 *
 * <code>
 * class MyComponent extends JSXComponent {
 *   render() {
 *     return <div>Hello World</div>
 *   }
 * }
 * </code>
 */
var JSXComponent = function (_Component) {
	_inherits(JSXComponent, _Component);

	function JSXComponent() {
		_classCallCheck(this, JSXComponent);

		return _possibleConstructorReturn(this, (JSXComponent.__proto__ || Object.getPrototypeOf(JSXComponent)).apply(this, arguments));
	}

	_createClass(JSXComponent, [{
		key: 'handleStateWillChange_',


		/**
   * Fires before state batch changes. Provides hook point for modifying
   *     state.
   * @param {Event} event
   * @protected
   */
		value: function handleStateWillChange_(event) {
			if (event.type !== 'state') {
				this.willReceiveProps(event.changes);
			}
		}

		/**
   * Returns props that are not used or declared in the component.
   * @return {Object} Object containing props
   */

	}, {
		key: 'otherProps',
		value: function otherProps() {
			var removeKeys = [].concat(_toConsumableArray(this.getDataManager().getPropsInstance(this).getStateKeys()), ['key', 'ref']);

			var retObj = _metal.object.mixin({}, this.props);

			for (var i = 0; i < removeKeys.length; i++) {
				var key = removeKeys[i];

				if (retObj.hasOwnProperty(key)) {
					delete retObj[key];
				}
			}

			return retObj;
		}

		/**
   * Lifecycle. Called when the component is about to receive new props.
   * Provides a hook point for modifying state that can be used in the next
   * rerender.
   * @param {Object} changes Changes made to this.props
   */

	}, {
		key: 'willReceiveProps',
		value: function willReceiveProps() {}
	}], [{
		key: 'render',

		/**
   * Creates and renders the given function, which can either be a simple
   * JSX function or a component constructor.
   * @param {!function()} fnOrCtor Either be a simple jsx dom function or a
   *     component constructor.
   * @param {Object=} opt_data Optional config data for the function.
   * @param {Element=} opt_element Optional parent for the rendered content.
   * @return {!Component} The rendered component's instance.
   * @override
   */
		value: function render() {
			return _metalIncrementalDom2.default.render.apply(_metalIncrementalDom2.default, arguments);
		}
	}]);

	return JSXComponent;
}(_metalComponent2.default);

JSXComponent.DATA_MANAGER = _JSXDataManager2.default;
JSXComponent.RENDERER = _JSXRenderer2.default;

exports.default = JSXComponent;
