// clay-autocomplete/lib/ClayAutocomplete.soy.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.templates = exports.ClayAutocomplete = undefined;

var _metalComponent = require('metal-component');

var _metalComponent2 = _interopRequireDefault(_metalComponent);

var _metalSoy = require('metal-soy');

var _metalSoy2 = _interopRequireDefault(_metalSoy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /* jshint ignore:start */


var templates;
goog.loadModule(function (exports) {
  var soy = goog.require('soy');
  var soydata = goog.require('soydata');
  // This file was automatically generated from ClayAutocomplete.soy.
  // Please don't edit this file by hand.

  /**
   * @fileoverview Templates in namespace ClayAutocomplete.
   * @hassoydeltemplate {ClayAutocomplete.Item.idom}
   * @hassoydeltemplate {ClayAutocomplete.ItemContent.idom}
   * @hassoydelcall {ClayAutocomplete.Item.idom}
   * @hassoydelcall {ClayAutocomplete.ItemContent.idom}
   * @public
   */

  goog.module('ClayAutocomplete.incrementaldom');

  goog.require('goog.soy.data.SanitizedContent');
  var incrementalDom = goog.require('incrementaldom');
  goog.require('soy');
  goog.require('soy.asserts');
  var soyIdom = goog.require('soy.idom');

  var $templateAlias1 = _metalSoy2.default.getTemplate('ClayDataProvider.incrementaldom', 'render');

  var $templateAlias2 = _metalSoy2.default.getTemplate('ClayPortal.incrementaldom', 'render');

  /**
   * @param {{
   *  dataSource: (!Array<?>|!Object|!goog.soy.data.SanitizedContent|*|string),
   *  _dropdownWidth: (null|number|undefined),
   *  _handleDataChange: (*|null|undefined),
   *  _handleDropdownItemClick: (*|null|undefined),
   *  _handleOnBlur: (*|null|undefined),
   *  _handleOnFocus: (*|null|undefined),
   *  _handleOnInput: (*|null|undefined),
   *  _handleOnKeydown: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  dataProviderInitialData: (!Array<?>|!Object|null|undefined),
   *  dropdownPortalElement: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  elementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  filteredItems: (!Array<?>|null|undefined),
   *  id: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputElementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  placeholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  pollingInterval: (null|number|undefined),
   *  requestInputMode: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  requestOptions: (null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}),
   *  requestRetries: (null|number|undefined),
   *  requestTimeout: (null|number|undefined),
   *  unstable_content: (function()|null|undefined),
   *  useDefaultClasses: (boolean|null|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $render(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    /** @type {!Array<?>|!Object|!goog.soy.data.SanitizedContent|*|string} */
    var dataSource = opt_data.dataSource;
    /** @type {null|number|undefined} */
    var _dropdownWidth = soy.asserts.assertType(opt_data._dropdownWidth == null || goog.isNumber(opt_data._dropdownWidth), '_dropdownWidth', opt_data._dropdownWidth, 'null|number|undefined');
    /** @type {*|null|undefined} */
    var _handleDataChange = opt_data._handleDataChange;
    /** @type {*|null|undefined} */
    var _handleDropdownItemClick = opt_data._handleDropdownItemClick;
    /** @type {*|null|undefined} */
    var _handleOnBlur = opt_data._handleOnBlur;
    /** @type {*|null|undefined} */
    var _handleOnFocus = opt_data._handleOnFocus;
    /** @type {*|null|undefined} */
    var _handleOnInput = opt_data._handleOnInput;
    /** @type {*|null|undefined} */
    var _handleOnKeydown = opt_data._handleOnKeydown;
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var contentRenderer = soy.asserts.assertType(opt_data.contentRenderer == null || goog.isString(opt_data.contentRenderer) || opt_data.contentRenderer instanceof goog.soy.data.SanitizedContent, 'contentRenderer', opt_data.contentRenderer, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!Array<?>|!Object|null|undefined} */
    var dataProviderInitialData = soy.asserts.assertType(opt_data.dataProviderInitialData == null || goog.isObject(opt_data.dataProviderInitialData) || goog.isArray(opt_data.dataProviderInitialData), 'dataProviderInitialData', opt_data.dataProviderInitialData, '!Array<?>|!Object|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var dropdownPortalElement = soy.asserts.assertType(opt_data.dropdownPortalElement == null || goog.isString(opt_data.dropdownPortalElement) || opt_data.dropdownPortalElement instanceof goog.soy.data.SanitizedContent, 'dropdownPortalElement', opt_data.dropdownPortalElement, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var elementClasses = soy.asserts.assertType(opt_data.elementClasses == null || goog.isString(opt_data.elementClasses) || opt_data.elementClasses instanceof goog.soy.data.SanitizedContent, 'elementClasses', opt_data.elementClasses, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!Array<?>|null|undefined} */
    var filteredItems = soy.asserts.assertType(opt_data.filteredItems == null || goog.isArray(opt_data.filteredItems), 'filteredItems', opt_data.filteredItems, '!Array<?>|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var id = soy.asserts.assertType(opt_data.id == null || goog.isString(opt_data.id) || opt_data.id instanceof goog.soy.data.SanitizedContent, 'id', opt_data.id, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var inputElementClasses = soy.asserts.assertType(opt_data.inputElementClasses == null || goog.isString(opt_data.inputElementClasses) || opt_data.inputElementClasses instanceof goog.soy.data.SanitizedContent, 'inputElementClasses', opt_data.inputElementClasses, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var inputName = soy.asserts.assertType(opt_data.inputName == null || goog.isString(opt_data.inputName) || opt_data.inputName instanceof goog.soy.data.SanitizedContent, 'inputName', opt_data.inputName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var inputValue = soy.asserts.assertType(opt_data.inputValue == null || goog.isString(opt_data.inputValue) || opt_data.inputValue instanceof goog.soy.data.SanitizedContent, 'inputValue', opt_data.inputValue, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var placeholder = soy.asserts.assertType(opt_data.placeholder == null || goog.isString(opt_data.placeholder) || opt_data.placeholder instanceof goog.soy.data.SanitizedContent, 'placeholder', opt_data.placeholder, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {null|number|undefined} */
    var pollingInterval = soy.asserts.assertType(opt_data.pollingInterval == null || goog.isNumber(opt_data.pollingInterval), 'pollingInterval', opt_data.pollingInterval, 'null|number|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var requestInputMode = soy.asserts.assertType(opt_data.requestInputMode == null || goog.isString(opt_data.requestInputMode) || opt_data.requestInputMode instanceof goog.soy.data.SanitizedContent, 'requestInputMode', opt_data.requestInputMode, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}} */
    var requestOptions = soy.asserts.assertType(opt_data.requestOptions == null || goog.isObject(opt_data.requestOptions), 'requestOptions', opt_data.requestOptions, 'null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}');
    /** @type {null|number|undefined} */
    var requestRetries = soy.asserts.assertType(opt_data.requestRetries == null || goog.isNumber(opt_data.requestRetries), 'requestRetries', opt_data.requestRetries, 'null|number|undefined');
    /** @type {null|number|undefined} */
    var requestTimeout = soy.asserts.assertType(opt_data.requestTimeout == null || goog.isNumber(opt_data.requestTimeout), 'requestTimeout', opt_data.requestTimeout, 'null|number|undefined');
    /** @type {function()|null|undefined} */
    var unstable_content = soy.asserts.assertType(opt_data.unstable_content == null || goog.isFunction(opt_data.unstable_content), 'unstable_content', opt_data.unstable_content, 'function()|null|undefined');
    /** @type {boolean|null|undefined} */
    var useDefaultClasses = soy.asserts.assertType(opt_data.useDefaultClasses == null || goog.isBoolean(opt_data.useDefaultClasses) || opt_data.useDefaultClasses === 1 || opt_data.useDefaultClasses === 0, 'useDefaultClasses', opt_data.useDefaultClasses, 'boolean|null|undefined');
    var attributes__soy29 = function attributes__soy29() {
      incrementalDom.attr('class', (elementClasses ? ' ' + elementClasses : '') + (useDefaultClasses != false ? ' dropdown-full form-group' : ''));
      if (id) {
        incrementalDom.attr('id', id);
      }
    };
    incrementalDom.elementOpenStart('div');
    attributes__soy29();
    incrementalDom.elementOpenEnd();
    soyIdom.print(unstable_content);
    $input(opt_data, null, opt_ijData);
    $dropdown(opt_data, null, opt_ijData);
    incrementalDom.elementClose('div');
  }
  exports.render = $render;
  /**
   * @typedef {{
   *  dataSource: (!Array<?>|!Object|!goog.soy.data.SanitizedContent|*|string),
   *  _dropdownWidth: (null|number|undefined),
   *  _handleDataChange: (*|null|undefined),
   *  _handleDropdownItemClick: (*|null|undefined),
   *  _handleOnBlur: (*|null|undefined),
   *  _handleOnFocus: (*|null|undefined),
   *  _handleOnInput: (*|null|undefined),
   *  _handleOnKeydown: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  dataProviderInitialData: (!Array<?>|!Object|null|undefined),
   *  dropdownPortalElement: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  elementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  filteredItems: (!Array<?>|null|undefined),
   *  id: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputElementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  placeholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  pollingInterval: (null|number|undefined),
   *  requestInputMode: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  requestOptions: (null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}),
   *  requestRetries: (null|number|undefined),
   *  requestTimeout: (null|number|undefined),
   *  unstable_content: (function()|null|undefined),
   *  useDefaultClasses: (boolean|null|undefined)
   * }}
   */
  $render.Params;
  if (goog.DEBUG) {
    $render.soyTemplateName = 'ClayAutocomplete.render';
  }

  /**
   * @param {{
   *  _handleOnBlur: (*|null|undefined),
   *  _handleOnFocus: (*|null|undefined),
   *  _handleOnInput: (*|null|undefined),
   *  _handleOnKeydown: (*|null|undefined),
   *  inputElementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  placeholder: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $input(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    opt_data = opt_data || {};
    /** @type {*|null|undefined} */
    var _handleOnBlur = opt_data._handleOnBlur;
    /** @type {*|null|undefined} */
    var _handleOnFocus = opt_data._handleOnFocus;
    /** @type {*|null|undefined} */
    var _handleOnInput = opt_data._handleOnInput;
    /** @type {*|null|undefined} */
    var _handleOnKeydown = opt_data._handleOnKeydown;
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var inputElementClasses = soy.asserts.assertType(opt_data.inputElementClasses == null || goog.isString(opt_data.inputElementClasses) || opt_data.inputElementClasses instanceof goog.soy.data.SanitizedContent, 'inputElementClasses', opt_data.inputElementClasses, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var inputName = soy.asserts.assertType(opt_data.inputName == null || goog.isString(opt_data.inputName) || opt_data.inputName instanceof goog.soy.data.SanitizedContent, 'inputName', opt_data.inputName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var inputValue = soy.asserts.assertType(opt_data.inputValue == null || goog.isString(opt_data.inputValue) || opt_data.inputValue instanceof goog.soy.data.SanitizedContent, 'inputValue', opt_data.inputValue, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var placeholder = soy.asserts.assertType(opt_data.placeholder == null || goog.isString(opt_data.placeholder) || opt_data.placeholder instanceof goog.soy.data.SanitizedContent, 'placeholder', opt_data.placeholder, '!goog.soy.data.SanitizedContent|null|string|undefined');
    var attributes__soy88 = function attributes__soy88() {
      incrementalDom.attr('class', inputElementClasses ? inputElementClasses : 'form-control');
      if (inputName) {
        incrementalDom.attr('id', inputName);
        incrementalDom.attr('name', inputName);
      }
      incrementalDom.attr('onBlur', _handleOnBlur);
      incrementalDom.attr('onFocus', _handleOnFocus);
      incrementalDom.attr('onInput', _handleOnInput);
      incrementalDom.attr('onKeydown', _handleOnKeydown);
      incrementalDom.attr('ref', 'input');
      incrementalDom.attr('type', 'text');
      if (placeholder) {
        incrementalDom.attr('placeholder', placeholder);
      }
      if (inputValue) {
        incrementalDom.attr('value', inputValue);
      } else {
        incrementalDom.attr('value', '');
      }
    };
    incrementalDom.elementOpenStart('input');
    attributes__soy88();
    incrementalDom.elementOpenEnd();
    incrementalDom.elementClose('input');
  }
  exports.input = $input;
  /**
   * @typedef {{
   *  _handleOnBlur: (*|null|undefined),
   *  _handleOnFocus: (*|null|undefined),
   *  _handleOnInput: (*|null|undefined),
   *  _handleOnKeydown: (*|null|undefined),
   *  inputElementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  inputValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  placeholder: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }}
   */
  $input.Params;
  if (goog.DEBUG) {
    $input.soyTemplateName = 'ClayAutocomplete.input';
  }

  /**
   * @param {{
   *  dataSource: (!Array<?>|!Object|!goog.soy.data.SanitizedContent|*|string),
   *  _dropdownWidth: (null|number|undefined),
   *  _handleDataChange: (*|null|undefined),
   *  _handleDropdownItemClick: (*|null|undefined),
   *  _handleOnKeydown: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  dataProviderInitialData: (!Array<?>|!Object|null|undefined),
   *  dropdownPortalElement: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  filteredItems: (!Array<?>|null|undefined),
   *  pollingInterval: (null|number|undefined),
   *  requestInputMode: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  requestOptions: (null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}),
   *  requestRetries: (null|number|undefined),
   *  requestTimeout: (null|number|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $dropdown(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    var $$temp;
    /** @type {!Array<?>|!Object|!goog.soy.data.SanitizedContent|*|string} */
    var dataSource = opt_data.dataSource;
    /** @type {null|number|undefined} */
    var _dropdownWidth = soy.asserts.assertType(opt_data._dropdownWidth == null || goog.isNumber(opt_data._dropdownWidth), '_dropdownWidth', opt_data._dropdownWidth, 'null|number|undefined');
    /** @type {*|null|undefined} */
    var _handleDataChange = opt_data._handleDataChange;
    /** @type {*|null|undefined} */
    var _handleDropdownItemClick = opt_data._handleDropdownItemClick;
    /** @type {*|null|undefined} */
    var _handleOnKeydown = opt_data._handleOnKeydown;
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var contentRenderer = soy.asserts.assertType(opt_data.contentRenderer == null || goog.isString(opt_data.contentRenderer) || opt_data.contentRenderer instanceof goog.soy.data.SanitizedContent, 'contentRenderer', opt_data.contentRenderer, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!Array<?>|!Object|null|undefined} */
    var dataProviderInitialData = soy.asserts.assertType(opt_data.dataProviderInitialData == null || goog.isObject(opt_data.dataProviderInitialData) || goog.isArray(opt_data.dataProviderInitialData), 'dataProviderInitialData', opt_data.dataProviderInitialData, '!Array<?>|!Object|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var dropdownPortalElement = soy.asserts.assertType(opt_data.dropdownPortalElement == null || goog.isString(opt_data.dropdownPortalElement) || opt_data.dropdownPortalElement instanceof goog.soy.data.SanitizedContent, 'dropdownPortalElement', opt_data.dropdownPortalElement, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!Array<?>|null|undefined} */
    var filteredItems = soy.asserts.assertType(opt_data.filteredItems == null || goog.isArray(opt_data.filteredItems), 'filteredItems', opt_data.filteredItems, '!Array<?>|null|undefined');
    /** @type {null|number|undefined} */
    var pollingInterval = soy.asserts.assertType(opt_data.pollingInterval == null || goog.isNumber(opt_data.pollingInterval), 'pollingInterval', opt_data.pollingInterval, 'null|number|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var requestInputMode = soy.asserts.assertType(opt_data.requestInputMode == null || goog.isString(opt_data.requestInputMode) || opt_data.requestInputMode instanceof goog.soy.data.SanitizedContent, 'requestInputMode', opt_data.requestInputMode, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}} */
    var requestOptions = soy.asserts.assertType(opt_data.requestOptions == null || goog.isObject(opt_data.requestOptions), 'requestOptions', opt_data.requestOptions, 'null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}');
    /** @type {null|number|undefined} */
    var requestRetries = soy.asserts.assertType(opt_data.requestRetries == null || goog.isNumber(opt_data.requestRetries), 'requestRetries', opt_data.requestRetries, 'null|number|undefined');
    /** @type {null|number|undefined} */
    var requestTimeout = soy.asserts.assertType(opt_data.requestTimeout == null || goog.isNumber(opt_data.requestTimeout), 'requestTimeout', opt_data.requestTimeout, 'null|number|undefined');
    var isVisible__soy147 = filteredItems && filteredItems.length > 0 ? true : false;
    var dropdownContent__soy149 = function dropdownContent__soy149() {
      var attributes__soy151 = function attributes__soy151() {
        incrementalDom.attr('class', 'autocomplete-dropdown-menu dropdown-menu' + (isVisible__soy147 ? ' show' : ''));
        incrementalDom.attr('data-onkeydown', _handleOnKeydown);
        if (_dropdownWidth) {
          incrementalDom.attr('style', 'width: ' + _dropdownWidth + 'px; max-width: none;');
        }
        incrementalDom.attr('ref', 'dropdown');
      };
      incrementalDom.elementOpenStart('ul');
      attributes__soy151();
      incrementalDom.elementOpenEnd();
      if (filteredItems) {
        var item178List = filteredItems;
        var item178ListLen = item178List.length;
        for (var item178Index = 0; item178Index < item178ListLen; item178Index++) {
          var item178Data = item178List[item178Index];
          soy.$$getDelegateFn(soy.$$getDelTemplateId('ClayAutocomplete.Item.idom'), contentRenderer, true)({ _handleDropdownItemClick: _handleDropdownItemClick, contentRenderer: contentRenderer, data: item178Data, index: item178Index }, null, opt_ijData);
        }
      }
      incrementalDom.elementClose('ul');
    };
    var param184 = function param184() {
      $templateAlias2({ content: dropdownContent__soy149, portalElement: ($$temp = dropdownPortalElement) == null ? '#clay_dropdown_portal' : $$temp, ref: 'portal' }, null, opt_ijData);
    };
    $templateAlias1({ content: param184, dataSource: dataSource, events: { dataChange: _handleDataChange }, initialData: dataProviderInitialData, pollingInterval: pollingInterval, ref: 'dataProvider', requestInputMode: requestInputMode, requestOptions: requestOptions, requestRetries: requestRetries, requestTimeout: requestTimeout }, null, opt_ijData);
  }
  exports.dropdown = $dropdown;
  /**
   * @typedef {{
   *  dataSource: (!Array<?>|!Object|!goog.soy.data.SanitizedContent|*|string),
   *  _dropdownWidth: (null|number|undefined),
   *  _handleDataChange: (*|null|undefined),
   *  _handleDropdownItemClick: (*|null|undefined),
   *  _handleOnKeydown: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  dataProviderInitialData: (!Array<?>|!Object|null|undefined),
   *  dropdownPortalElement: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  filteredItems: (!Array<?>|null|undefined),
   *  pollingInterval: (null|number|undefined),
   *  requestInputMode: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  requestOptions: (null|undefined|{body: !Object, cache: (!goog.soy.data.SanitizedContent|string), credentials: (!goog.soy.data.SanitizedContent|string), headers: !Object, method: (!goog.soy.data.SanitizedContent|string), mode: (!goog.soy.data.SanitizedContent|string), redirect: (!goog.soy.data.SanitizedContent|string), referrer: (!goog.soy.data.SanitizedContent|string)}),
   *  requestRetries: (null|number|undefined),
   *  requestTimeout: (null|number|undefined)
   * }}
   */
  $dropdown.Params;
  if (goog.DEBUG) {
    $dropdown.soyTemplateName = 'ClayAutocomplete.dropdown';
  }

  /**
   * @param {{
   *  data: (?),
   *  index: number,
   *  _handleDropdownItemClick: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function __deltemplate_s201_4a91529a(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    /** @type {?} */
    var data = opt_data.data;
    /** @type {number} */
    var index = soy.asserts.assertType(goog.isNumber(opt_data.index), 'index', opt_data.index, 'number');
    /** @type {*|null|undefined} */
    var _handleDropdownItemClick = opt_data._handleDropdownItemClick;
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var contentRenderer = soy.asserts.assertType(opt_data.contentRenderer == null || goog.isString(opt_data.contentRenderer) || opt_data.contentRenderer instanceof goog.soy.data.SanitizedContent, 'contentRenderer', opt_data.contentRenderer, '!goog.soy.data.SanitizedContent|null|string|undefined');
    incrementalDom.elementOpen('li');
    incrementalDom.elementOpenStart('a');
    incrementalDom.attr('data-onclick', _handleDropdownItemClick);
    incrementalDom.attr('data-dropdown-item-index', index);
    incrementalDom.attr('id', 'item');
    incrementalDom.attr('class', 'dropdown-item');
    incrementalDom.attr('href', 'javascript:;');
    incrementalDom.elementOpenEnd();
    soy.$$getDelegateFn(soy.$$getDelTemplateId('ClayAutocomplete.ItemContent.idom'), contentRenderer, true)(opt_data, null, opt_ijData);
    incrementalDom.elementClose('a');
    incrementalDom.elementClose('li');
  }
  exports.__deltemplate_s201_4a91529a = __deltemplate_s201_4a91529a;
  /**
   * @typedef {{
   *  data: (?),
   *  index: number,
   *  _handleDropdownItemClick: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }}
   */
  __deltemplate_s201_4a91529a.Params;
  if (goog.DEBUG) {
    __deltemplate_s201_4a91529a.soyTemplateName = 'ClayAutocomplete.__deltemplate_s201_4a91529a';
  }
  soy.$$registerDelegateFn(soy.$$getDelTemplateId('ClayAutocomplete.Item.idom'), '', 0, __deltemplate_s201_4a91529a);

  /**
   * @param {{
   *  data: (?)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function __deltemplate_s215_40e63d57(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    /** @type {?} */
    var data = opt_data.data;
    if (data.matches && data.matches.length > 0) {
      var char232List = data.matches;
      var char232ListLen = char232List.length;
      for (var char232Index = 0; char232Index < char232ListLen; char232Index++) {
        var char232Data = char232List[char232Index];
        if (char232Data.match) {
          incrementalDom.elementOpen('strong');
          soyIdom.print(char232Data.value);
          incrementalDom.elementClose('strong');
        } else {
          soyIdom.print(char232Data.value);
        }
      }
    }
  }
  exports.__deltemplate_s215_40e63d57 = __deltemplate_s215_40e63d57;
  /**
   * @typedef {{
   *  data: (?)
   * }}
   */
  __deltemplate_s215_40e63d57.Params;
  if (goog.DEBUG) {
    __deltemplate_s215_40e63d57.soyTemplateName = 'ClayAutocomplete.__deltemplate_s215_40e63d57';
  }
  soy.$$registerDelegateFn(soy.$$getDelTemplateId('ClayAutocomplete.ItemContent.idom'), '', 0, __deltemplate_s215_40e63d57);

  exports.render.params = ["dataSource", "_dropdownWidth", "_handleDataChange", "_handleDropdownItemClick", "_handleOnBlur", "_handleOnFocus", "_handleOnInput", "_handleOnKeydown", "contentRenderer", "dataProviderInitialData", "dropdownPortalElement", "elementClasses", "filteredItems", "id", "inputElementClasses", "inputName", "inputValue", "placeholder", "pollingInterval", "requestInputMode", "requestOptions", "requestRetries", "requestTimeout", "unstable_content", "useDefaultClasses"];
  exports.render.types = { "dataSource": "string|[]|list<?>|any", "_dropdownWidth": "number", "_handleDataChange": "any", "_handleDropdownItemClick": "any", "_handleOnBlur": "any", "_handleOnFocus": "any", "_handleOnInput": "any", "_handleOnKeydown": "any", "contentRenderer": "string", "dataProviderInitialData": "[]|list<?>", "dropdownPortalElement": "string", "elementClasses": "string", "filteredItems": "list<?>", "id": "string", "inputElementClasses": "string", "inputName": "string", "inputValue": "string", "placeholder": "string", "pollingInterval": "number", "requestInputMode": "string", "requestOptions": "[\n\t\tmethod: string,\n\t\tmode: string,\n\t\tcache: string,\n\t\tcredentials: string,\n\t\theaders: [],\n\t\tredirect: string,\n\t\treferrer: string,\n\t\tbody: []\n\t]", "requestRetries": "number", "requestTimeout": "number", "unstable_content": "html", "useDefaultClasses": "bool" };
  exports.input.params = ["_handleOnBlur", "_handleOnFocus", "_handleOnInput", "_handleOnKeydown", "inputElementClasses", "inputName", "inputValue", "placeholder"];
  exports.input.types = { "_handleOnBlur": "any", "_handleOnFocus": "any", "_handleOnInput": "any", "_handleOnKeydown": "any", "inputElementClasses": "string", "inputName": "string", "inputValue": "string", "placeholder": "string" };
  exports.dropdown.params = ["dataSource", "_dropdownWidth", "_handleDataChange", "_handleDropdownItemClick", "_handleOnKeydown", "contentRenderer", "dataProviderInitialData", "dropdownPortalElement", "filteredItems", "pollingInterval", "requestInputMode", "requestOptions", "requestRetries", "requestTimeout"];
  exports.dropdown.types = { "dataSource": "string|[]|list<?>|any", "_dropdownWidth": "number", "_handleDataChange": "any", "_handleDropdownItemClick": "any", "_handleOnKeydown": "any", "contentRenderer": "string", "dataProviderInitialData": "[]|list<?>", "dropdownPortalElement": "string", "filteredItems": "list<?>", "pollingInterval": "number", "requestInputMode": "string", "requestOptions": "[\n\t\tmethod: string,\n\t\tmode: string,\n\t\tcache: string,\n\t\tcredentials: string,\n\t\theaders: [],\n\t\tredirect: string,\n\t\treferrer: string,\n\t\tbody: []\n\t]", "requestRetries": "number", "requestTimeout": "number" };
  exports.templates = templates = exports;
  return exports;
});

var ClayAutocomplete = function (_Component) {
  _inherits(ClayAutocomplete, _Component);

  function ClayAutocomplete() {
    _classCallCheck(this, ClayAutocomplete);

    return _possibleConstructorReturn(this, (ClayAutocomplete.__proto__ || Object.getPrototypeOf(ClayAutocomplete)).apply(this, arguments));
  }

  return ClayAutocomplete;
}(_metalComponent2.default);

_metalSoy2.default.register(ClayAutocomplete, templates);
exports.ClayAutocomplete = ClayAutocomplete;
exports.templates = templates;
exports.default = templates;
/* jshint ignore:end */
