// metal-key/lib/keyConstants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var BACKSPACE = exports.BACKSPACE = 8;
var TAB = exports.TAB = 9;
var NUM_CENTER = exports.NUM_CENTER = 12;
var ENTER = exports.ENTER = 13;
var RETURN = exports.RETURN = 13;
var SHIFT = exports.SHIFT = 16;
var CTRL = exports.CTRL = 17;
var ALT = exports.ALT = 18;
var PAUSE = exports.PAUSE = 19;
var CAPS_LOCK = exports.CAPS_LOCK = 20;
var ESC = exports.ESC = 27;
var SPACE = exports.SPACE = 32;
var PAGE_UP = exports.PAGE_UP = 33;
var PAGE_DOWN = exports.PAGE_DOWN = 34;
var END = exports.END = 35;
var HOME = exports.HOME = 36;
var LEFT = exports.LEFT = 37;
var UP = exports.UP = 38;
var RIGHT = exports.RIGHT = 39;
var DOWN = exports.DOWN = 40;
var PRINT_SCREEN = exports.PRINT_SCREEN = 44;
var INSERT = exports.INSERT = 45;
var DELETE = exports.DELETE = 46;
var ZERO = exports.ZERO = 48;
var ONE = exports.ONE = 49;
var TWO = exports.TWO = 50;
var THREE = exports.THREE = 51;
var FOUR = exports.FOUR = 52;
var FIVE = exports.FIVE = 53;
var SIX = exports.SIX = 54;
var SEVEN = exports.SEVEN = 55;
var EIGHT = exports.EIGHT = 56;
var NINE = exports.NINE = 57;
var A = exports.A = 65;
var B = exports.B = 66;
var C = exports.C = 67;
var D = exports.D = 68;
var E = exports.E = 69;
var F = exports.F = 70;
var G = exports.G = 71;
var H = exports.H = 72;
var I = exports.I = 73;
var J = exports.J = 74;
var K = exports.K = 75;
var L = exports.L = 76;
var M = exports.M = 77;
var N = exports.N = 78;
var O = exports.O = 79;
var P = exports.P = 80;
var Q = exports.Q = 81;
var R = exports.R = 82;
var S = exports.S = 83;
var T = exports.T = 84;
var U = exports.U = 85;
var V = exports.V = 86;
var W = exports.W = 87;
var X = exports.X = 88;
var Y = exports.Y = 89;
var Z = exports.Z = 90;
var CONTEXT_MENU = exports.CONTEXT_MENU = 93;
var NUM_ZERO = exports.NUM_ZERO = 96;
var NUM_ONE = exports.NUM_ONE = 97;
var NUM_TWO = exports.NUM_TWO = 98;
var NUM_THREE = exports.NUM_THREE = 99;
var NUM_FOUR = exports.NUM_FOUR = 100;
var NUM_FIVE = exports.NUM_FIVE = 101;
var NUM_SIX = exports.NUM_SIX = 102;
var NUM_SEVEN = exports.NUM_SEVEN = 103;
var NUM_EIGHT = exports.NUM_EIGHT = 104;
var NUM_NINE = exports.NUM_NINE = 105;
var NUM_MULTIPLY = exports.NUM_MULTIPLY = 106;
var NUM_PLUS = exports.NUM_PLUS = 107;
var NUM_MINUS = exports.NUM_MINUS = 109;
var NUM_PERIOD = exports.NUM_PERIOD = 110;
var NUM_DIVISION = exports.NUM_DIVISION = 111;
var F1 = exports.F1 = 112;
var F2 = exports.F2 = 113;
var F3 = exports.F3 = 114;
var F4 = exports.F4 = 115;
var F5 = exports.F5 = 116;
var F6 = exports.F6 = 117;
var F7 = exports.F7 = 118;
var F8 = exports.F8 = 119;
var F9 = exports.F9 = 120;
var F10 = exports.F10 = 121;
var F11 = exports.F11 = 122;
var F12 = exports.F12 = 123;
