// clay-list/lib/ClayList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.ClayList = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

require('clay-checkbox');

var _clayDropdown = require('clay-dropdown');

require('clay-icon');

require('clay-label');

require('clay-link');

require('clay-sticker');

var _metalState = require('metal-state');

var _clayComponent = require('clay-component');

var _clayComponent2 = _interopRequireDefault(_clayComponent);

var _metalWebComponent = require('metal-web-component');

var _metalWebComponent2 = _interopRequireDefault(_metalWebComponent);

var _metalSoy = require('metal-soy');

var _metalSoy2 = _interopRequireDefault(_metalSoy);

var _ClayListSoy = require('./ClayList.soy.js');

var _ClayListSoy2 = _interopRequireDefault(_ClayListSoy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
// eslint-disable-next-line


/**
 * Metal ClayList component.
 * @extends ClayComponent
 */
var ClayList = function (_ClayComponent) {
	_inherits(ClayList, _ClayComponent);

	function ClayList() {
		_classCallCheck(this, ClayList);

		return _possibleConstructorReturn(this, (ClayList.__proto__ || Object.getPrototypeOf(ClayList)).apply(this, arguments));
	}

	_createClass(ClayList, [{
		key: '_handleItemToggled',

		/**
   * Continues the propagation of the checkbox changed event
   * @param {!Event} event
   * @private
   */
		value: function _handleItemToggled(event) {
			this.emit('itemToggled', event);
		}
	}]);

	return ClayList;
}(_clayComponent2.default);

/**
 * State definition.
 * @static
 * @type {!Object}
 */


ClayList.STATE = {
	/**
  * Data to add to the element.
  * @default undefined
  * @instance
  * @memberof ClayList
  * @type {?object}
  */
	data: _metalState.Config.object(),

	/**
  * Object that wires events with default listeners
  * @default undefined
  * @instance
  * @memberof ClayList
  * @review
  * @type {?(object|undefined)}
  */
	defaultEventHandler: _metalState.Config.object(),

	/**
  * CSS classes to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayList
  * @type {?(string|undefined)}
  */
	elementClasses: _metalState.Config.string(),

	/**
  * Id to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayList
  * @type {?(string|undefined)}
  */
	id: _metalState.Config.string(),

	/**
  * List of items.
  * @default undefined
  * @instance
  * @memberof ClayList
  * @type {?(array|undefined)}
  */
	items: _metalState.Config.array(),

	/**
  * Flag to indicate if the list group items are selectable.
  * @default false
  * @instance
  * @memberof ClayList
  * @type {?(bool|undefined)}
  */
	selectable: _metalState.Config.bool().value(false),

	/**
  * Schema mapping list item fields with item data properties.
  * @default undefined
  * @instance
  * @memberof ClayList
  * @type {!object}
  */
	schema: _metalState.Config.shapeOf({
		fields: _metalState.Config.arrayOf(_metalState.Config.shapeOf({
			contentRenderer: _metalState.Config.string(),
			contentRendererMap: _metalState.Config.object(),
			fieldName: _metalState.Config.string(),
			fieldsMap: _metalState.Config.object(),
			iconsMap: _metalState.Config.object(),
			iconShapesMap: _metalState.Config.object(),
			labelStylesMap: _metalState.Config.object()
		})),
		inputNameField: _metalState.Config.string(),
		inputNamesMap: _metalState.Config.object(),
		inputValueField: _metalState.Config.string()
	}).required(),

	/**
  * The path to the SVG spritemap file containing the icons.
  * @default undefined
  * @instance
  * @memberof ClayList
  * @type {?(string|undefined)}
  */
	spritemap: _metalState.Config.string()
};

(0, _metalWebComponent2.default)('clay-list', ClayList);

_metalSoy2.default.register(ClayList, _ClayListSoy2.default);

exports.ClayList = ClayList;
exports.default = ClayList;
