// clay-badge/lib/ClayBadge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.ClayBadge = undefined;

var _clayComponent = require('clay-component');

var _clayComponent2 = _interopRequireDefault(_clayComponent);

var _metalWebComponent = require('metal-web-component');

var _metalWebComponent2 = _interopRequireDefault(_metalWebComponent);

var _metalSoy = require('metal-soy');

var _metalSoy2 = _interopRequireDefault(_metalSoy);

var _metalState = require('metal-state');

var _ClayBadgeSoy = require('./ClayBadge.soy.js');

var _ClayBadgeSoy2 = _interopRequireDefault(_ClayBadgeSoy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Metal Clay Badge component.
 * @extends ClayComponent
 */
var ClayBadge = function (_ClayComponent) {
	_inherits(ClayBadge, _ClayComponent);

	function ClayBadge() {
		_classCallCheck(this, ClayBadge);

		return _possibleConstructorReturn(this, (ClayBadge.__proto__ || Object.getPrototypeOf(ClayBadge)).apply(this, arguments));
	}

	return ClayBadge;
}(_clayComponent2.default);

/**
 * State definition.
 * @static
 * @type {!Object}
 */


ClayBadge.STATE = {
	/**
  * Data to add to the element.
  * @default undefined
  * @instance
  * @memberof ClayBadge
  * @type {?object}
  */
	data: _metalState.Config.object(),

	/**
  * Object that wires events with default listeners
  * @default undefined
  * @instance
  * @memberof ClayBadge
  * @review
  * @type {?(object|undefined)}
  */
	defaultEventHandler: _metalState.Config.object(),

	/**
  * CSS classes to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayBadge
  * @type {?(string|undefined)}
  */
	elementClasses: _metalState.Config.string(),

	/**
  * Id to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayBadge
  * @type {?(string|undefined)}
  */
	id: _metalState.Config.string(),

	/**
  * The label of the badge element.
  * @default undefined
  * @instance
  * @memberof ClayBadge
  * @type {!string}
  */
	label: _metalState.Config.string().required(),

	/**
  * Badge style color.
  * @default primary
  * @instance
  * @memberof ClayBadge
  * @type {?string}
  */
	style: _metalState.Config.oneOf(['danger', 'info', 'primary', 'secondary', 'success', 'warning']).value('primary')
};

(0, _metalWebComponent2.default)('clay-badge', ClayBadge);

_metalSoy2.default.register(ClayBadge, _ClayBadgeSoy2.default);

exports.ClayBadge = ClayBadge;
exports.default = ClayBadge;
