// liferay-npm-bundler/lib/config/internal/bundler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.init = init;
exports.getExclusions = getExclusions;
exports.getIncludeDependencies = getIncludeDependencies;
exports.getPlugins = getPlugins;
exports.getMaxParallelFiles = getMaxParallelFiles;

var _dotProp = require('dot-prop');

var _dotProp2 = _interopRequireDefault(_dotProp);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * © 2017 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

var config = void 0;

/**
 * Initialize submodule
 * @param {object} state
 * @return {void}
 */
function init(state) {
	config = state.config;
}

/**
 * Get the configured file exclusions for a given package.
 * @param {PkgDesc} pkg the package descriptor
 * @return {Array} an array of glob expressions
 */
function getExclusions(pkg) {
	var exclusions = config.exclude || {};

	// If it is explicitly false, return an empty exclusions array
	if (exclusions[pkg.id] === false || exclusions[pkg.name] === false || exclusions['*'] === false) {
		return [];
	}

	// If it is explicitly true, return an array with '**/*'
	if (exclusions[pkg.id] === true || exclusions[pkg.name] === true || exclusions['*'] === true) {
		return ['**/*'];
	}

	// In any other case, return what's in the config
	exclusions = exclusions[pkg.id] || exclusions[pkg.name] || exclusions['*'] || [];

	return exclusions;
}

/**
 * Extra dependencies to add to the final bundle (in addition to those listed
 * under the dependencies section of package.json).
 * @return {Array} an array of package names
 */
function getIncludeDependencies() {
	return _dotProp2.default.get(config, 'include-dependencies', []);
}

/**
 * Get the liferay-nmp-bundler plugins for a given package.
 * @param {String} phase 'pre', 'post' or 'copy'
 * @param {PkgDesc} pkg the package descriptor
 * @return {Array} the instantiated Babel plugins
 */
function getPlugins(phase, pkg) {
	var pluginsKey = {
		copy: 'copy-plugins',
		pre: 'plugins',
		post: 'post-plugins'
	}[phase];

	var pluginNames = (0, _util.getPackageConfig)(pkg, pluginsKey, []);

	return instantiatePlugins(pluginNames);
}

/**
 * Get maximum number of files to process in parallel.
 * @return {number}
 */
function getMaxParallelFiles() {
	// Default values for "ulimit -n" vary across different OSes. Some values
	// I have found are:
	//   - Apparently Mac OS X limit is 256 but usually people increase it
	//   - Fedora: 1024
	//   - Windows: there's no ulimit, but MSVCRT.DLL has a 2048 limit
	// Given this mess and the impossibility of retrieving the limit from Node,
	// I'm giving this a default value of 128 because it looks like it doesn't
	// impact performance and should be low enough to make it work in all OSes.
	return _dotProp2.default.get(config, 'max-parallel-files', 128);
}

/**
 * Instantiate bundler plugins described by their names.
 * @param  {Array} pluginNames list of plugin names to instantiate
 * @return {Array} list of plugin descriptors with name, config and run fields
 */
function instantiatePlugins(pluginNames) {
	return pluginNames.map(function (pluginName) {
		var pluginConfig = {};

		if (Array.isArray(pluginName)) {
			pluginConfig = pluginName[1];
			pluginName = pluginName[0];
		}

		var pluginModule = (0, _util.configRequire)('liferay-npm-bundler-plugin-' + pluginName);

		return {
			name: pluginName,
			config: pluginConfig,
			run: pluginModule.default
		};
	});
}
