// clay-radio/lib/ClayRadio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.ClayRadio = undefined;

var _clayComponent = require('clay-component');

var _clayComponent2 = _interopRequireDefault(_clayComponent);

var _metalWebComponent = require('metal-web-component');

var _metalWebComponent2 = _interopRequireDefault(_metalWebComponent);

var _metalSoy = require('metal-soy');

var _metalSoy2 = _interopRequireDefault(_metalSoy);

var _metalState = require('metal-state');

var _ClayRadioSoy = require('./ClayRadio.soy.js');

var _ClayRadioSoy2 = _interopRequireDefault(_ClayRadioSoy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Implementation of the Metal Clay Radio.
 * @extends ClayComponent
 */
var ClayRadio = function (_ClayComponent) {
	_inherits(ClayRadio, _ClayComponent);

	function ClayRadio() {
		_classCallCheck(this, ClayRadio);

		return _possibleConstructorReturn(this, (ClayRadio.__proto__ || Object.getPrototypeOf(ClayRadio)).apply(this, arguments));
	}

	return ClayRadio;
}(_clayComponent2.default);

/**
 * State definition.
 * @type {!Object}
 * @static
 */


ClayRadio.STATE = {
	/**
  * Flag to indicate if checkbox is checked.
  * @default false
  * @instance
  * @memberof ClayRadio
  * @type {?bool}
  */
	checked: _metalState.Config.bool().value(false),

	/**
  * Data to add to the element.
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @type {?object}
  */
	data: _metalState.Config.object(),

	/**
  * Object that wires events with default listeners
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @review
  * @type {?(object|undefined)}
  */
	defaultEventHandler: _metalState.Config.object(),

	/**
  * Flag to indicate if checkbox is disabled.
  * @default false
  * @instance
  * @memberof ClayRadio
  * @type {?bool}
  */
	disabled: _metalState.Config.bool().value(false),

	/**
  * CSS classes to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @type {?(string|undefined)}
  */
	elementClasses: _metalState.Config.string(),

	/**
  * Id to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @type {?(string|undefined)}
  */
	id: _metalState.Config.string(),

	/**
  * Flag to indicate if input should be wrapped to be inline or not.
  * @default false
  * @instance
  * @memberof ClayRadio
  * @type {?bool}
  */
	inline: _metalState.Config.bool().value(false),

	/**
  * Label of the input. Required for accessibility.
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @type {?(string|undefined)}
  */
	label: _metalState.Config.string(),

	/**
  * Custom content of the radio label. Use it to use your custom html.
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @type {?(html|undefined)}
  */
	labelContent: _metalState.Config.any(),

	/**
  * Name to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @type {?(string|undefined)}
  */
	name: _metalState.Config.string(),

	/**
  * Flag to indicate if label should be hide.
  * @default true
  * @instance
  * @memberof ClayRadio
  * @type {?bool}
  */
	showLabel: _metalState.Config.bool().value(true),

	/**
  * Value to be applied to the element.
  * @default undefined
  * @instance
  * @memberof ClayRadio
  * @type {?(string|undefined)}
  */
	value: _metalState.Config.string()
};

(0, _metalWebComponent2.default)('clay-radio', ClayRadio);

_metalSoy2.default.register(ClayRadio, _ClayRadioSoy2.default);

exports.ClayRadio = ClayRadio;
exports.default = ClayRadio;
